src/WebJobs.Script/Workers/WorkerConcurrencyManager.cs

```diff
@@ -0,0 +1,218 @@
+            {
+                // don't allow background exceptions to escape
+                _logger.LogError(ex.ToString());
+            }
+            _timer.Start();
+        }
+
+        internal bool AddWorkerIfNeeded(IDictionary<string, WorkerStatus> workerStatuses, TimeSpan elaspsedFromLastAdding)
+        {
+            if (elaspsedFromLastAdding < _concurrencyOptions.Value.AdjustmentPeriod)
+            {
+                return false;
+            }
+
+            bool result = false;
+            if (workerStatuses.All(x => x.Value.IsReady))
+            {
+                // Check how many channels are oveloaded
+                List<WorkerDescription> descriptions = new List<WorkerDescription>();
+                foreach (string key in workerStatuses.Keys)
+                {
+                    WorkerStatus workerStatus = workerStatuses[key];
+                    bool overloaded = IsOverloaded(workerStatus);
+                    descriptions.Add(new WorkerDescription()
+                    {
+                        WorkerId = key,
+                        WorkerStatus = workerStatus,
+                        Overloaded = overloaded
+                    });
+                }
+
+                int overloadedCount = descriptions.Where(x => x.Overloaded == true).Count();
+                if (overloadedCount > 0)
+                {
+                    if (workerStatuses.Count() < _concurrencyOptions.Value.MaxWorkerCount)
+                    {
+                        _logger.LogDebug($"Adding a new worker, overloaded workers = {overloadedCount}, initialized workers = {workerStatuses.Count()} ");
+                        result = true;
+                    }
+                }
+
+                if (result == true || _logStateStopWatch.Elapsed > _logStateInterval)
+                {
+                    StringBuilder sb = new StringBuilder();
+                    foreach (WorkerDescription description in descriptions)
+                    {
+                        sb.Append(LogWorkerState(description));
+                        sb.Append(Environment.NewLine);
+                    }
+                    _logStateStopWatch.Restart();
+                    _logger.LogDebug(sb.ToString());
+                }
+            }
+
+            return result;
+        }
+
+        internal bool IsOverloaded(WorkerStatus status)
+        {
+            if (status.WorkerStats.LatencyHistory.Count() >= _concurrencyOptions.Value.HistorySize)
+            {
+                int overloadedCount = status.WorkerStats.LatencyHistory.Where(x => x.TotalMilliseconds >= _concurrencyOptions.Value.LatencyThreshold.TotalMilliseconds).Count();
+                double proportion = (double)overloadedCount / _concurrencyOptions.Value.HistorySize;
+
+                return proportion >= _concurrencyOptions.Value.HistoryThreshold;
+            }
+            return false;
+        }
+
+        internal string LogWorkerState(WorkerDescription desc)
```

Comment:
rename to FormatWorkerDescription since you're not logging here. Better yet, make this a Format method on WorkerDescription itself.